sdks/java/io/singlestore/src/main/java/org/apache/beam/sdk/io/singlestore/SingleStoreDefaultUserDataMapper.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.singlestore;
+
+import java.math.BigDecimal;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.values.Row;
+import org.joda.time.Instant;
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+
+final class SingleStoreDefaultUserDataMapper implements SingleStoreIO.UserDataMapper<Row> {
+
+  private final transient DateTimeFormatter formatter =
+      DateTimeFormat.forPattern("yyyy-MM-DD' 'HH:mm:ss.SSS");
+
+  private String convertLogicalTypeFieldToString(Schema.FieldType type, Object value) {
+    Schema.LogicalType<Object, Object> logicalType =
+        (Schema.LogicalType<Object, Object>) type.getLogicalType();
+    if (logicalType == null) {
+      throw new UnsupportedOperationException("Failed to extract logical type");
+    }
```

Comment:
You can make use of [`FieldType::isLogicalType`](https://github.com/apache/beam/blob/3ffaa49c3217386291fa163d0a868a686b329ac6/sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java#L643) as a check here.